bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ReadOnlyLedgerHandle.java

```diff
@@ -207,4 +206,197 @@ public void readLastConfirmedComplete(int rc, long lastConfirmed, Object ctx) {
+        List<BookieSocketAddress> newEnsemble = new ArrayList<>(oldEnsemble);
+
+        int ensembleSize = metadata.getEnsembleSize();
+        int writeQ = metadata.getWriteQuorumSize();
+        int ackQ = metadata.getAckQuorumSize();
+        Map<String, byte[]> customMetadata = metadata.getCustomMetadata();
+
+        Set<BookieSocketAddress> exclude = new HashSet<>(failedBookies.values());
+
+        int replaced = 0;
+        for (Map.Entry<Integer, BookieSocketAddress> entry : failedBookies.entrySet()) {
+            int idx = entry.getKey();
+            BookieSocketAddress addr = entry.getValue();
+            LOG.debug("[EnsembleChange-L{}] replacing bookie: {} index: {}", getId(), addr, idx);
+
+            if (!newEnsemble.get(idx).equals(addr)) {
+                LOG.debug("[EnsembleChange-L{}] Not changing failed bookie {} at index {}, already changed to {}",
+                          getId(), addr, idx, newEnsemble.get(idx));
+                continue;
+            }
+            try {
+                BookieSocketAddress newBookie = clientCtx.getBookieWatcher().replaceBookie(
+                        ensembleSize, writeQ, ackQ, customMetadata, newEnsemble, idx, exclude);
+                newEnsemble.set(idx, newBookie);
+
+                replaced++;
+            } catch (BKException.BKNotEnoughBookiesException e) {
+                // if there is no bookie replaced, we throw not enough bookie exception
+                if (replaced <= 0) {
+                    throw e;
+                } else {
+                    break;
+                }
+            }
+        }
+        return newEnsemble;
+    }
+
+    private static Set<Integer> diffEnsemble(List<BookieSocketAddress> e1,
+                                             List<BookieSocketAddress> e2) {
+        checkArgument(e1.size() == e2.size(), "Ensembles must be of same size");
+        Set<Integer> diff = new HashSet<>();
+        for (int i = 0; i < e1.size(); i++) {
+            if (!e1.get(i).equals(e2.get(i))) {
+                diff.add(i);
+            }
+        }
+        return diff;
+    }
+
+    /**
+     * For a read only ledger handle, this method will only ever be called during recovery,
+     * when we are reading forward from LAC and writing back those entries. As such,
+     * unlike with LedgerHandle, we do not want to persist changes to the metadata as they occur,
+     * but rather, we want to defer the persistence until recovery has completed, and do it all
+     * on the close.
+     */
+    @Override
+    void handleBookieFailure(final Map<Integer, BookieSocketAddress> failedBookies) {
+        blockAddCompletions.incrementAndGet();
+
+        // handleBookieFailure should always run in the ordered executor thread for this
+        // ledger, so this synchronized should be unnecessary, but putting it here now
+        // just in case (can be removed when we validate threads)
+        synchronized (this) {
```

Comment:
this seems to change to a broader lock scope. I would suggest to have a metadataLock for locking metadata during this refactor period, this would reduce the risk of unknowns. we can consider removing the metadata lock once you done with immutable ledger metadata change.

Reply:
will create the lock for locking this and the newEnsembles.